torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+                # NOTE: There're certain cases where the placeholder nodes do
+                # not have real tensor values:
+                # 1. optimizer states can be None sometimes, i.e. SGD with
+                #    no momentum, optimizer states populate `momentum` state
+                #    as None, the full graph we get from `compile` would have
+                #    None as the placeholder value
+                # 2. function args might not only contain params or activations,
+                #    but also contain other non-tensor inputs, i.e. the model
+                #    and optimizer instances baked in as a placeholder, there might
+                #    also be some scalar argument which is not a tensor
+                #
+                # For the above cases, we create a NON_TENSOR stratgy so that we
+                # know it's not a tensor and we don't need to shard it
+                dp_strategy_map[node] = DataParallelStrategy(NodeType.NON_TENSOR, [])
+
+            elif placeholder_idx < num_params:
+                # during compilation there's an assumption that the first num_params
+                # placeholders should be parameters
+                shard_strategy = _gen_shard_strategy(mesh, 0)
+                replica_strategy = _gen_replicate_strategy(mesh)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.PARAM, [replica_strategy, shard_strategy]
+                )
+
+            elif placeholder_idx < activation_idx:
+                # optimizer states follow the same strategy as
+                # the corresponding parameters
+                placemenet_strategies: List[PlacementStrategy] = []
+
+                replica_strategy = _gen_replicate_strategy(mesh)
+                placemenet_strategies.append(replica_strategy)
+
+                if node.meta["val"].ndim > 0:
+                    # if the state is a scalar, we don't want to shard it
+                    # i.e. param state step is usually a scalar tensor
+                    shard_strategy = _gen_shard_strategy(mesh, 0)
+                    placemenet_strategies.append(shard_strategy)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.STATE, placemenet_strategies
+                )
+            else:
+                activation_batch_dim_size = node.meta["val"].shape[batch_dim]
+                # find the first activation node and use its batch dim size
+                if batch_dim_analzer.batch_dim_size == -1:
+                    batch_dim_analzer.init_batch_dim_size(activation_batch_dim_size)
+
+                batch_dim_analzer.set_batch_dim(node, batch_dim)
+                shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.ACT, [shard_strategy]
+                )
+            placeholder_idx += 1
+        elif node.op == "call_function":
```

Comment:
Meta comment: the propagation logic here is very long. It feels like there should be some way to make it shorter

Reply:
The length makes it difficult to tell which rules are "special cased" for FSDP, versus which rules would have worked in a more generic DTensor propagation environment

Reply:
Thanks let me try and see how to split this in a cleaner way!